audio/out/buffer.c

```diff
@@ -355,6 +355,11 @@ void ao_set_paused(struct ao *ao, bool paused)
     bool wakeup = false;
     bool do_reset = false, do_start = false;
 
+    // If we are going to pause and ao is still playing,
+    // be sure to drain the ao first for gapless.
+    if (paused && ao_is_playing(ao))
+        ao_drain(ao);
```

Comment:
hmm, I'm wondering if draining on every pause isn't too excessive or has negative side-effects.
non-gapless playback doesn't need this, right?

Reply:
For non-gapless, ao_is_playing is false. But this should only need to run on the end of file not every single pause. Let me see if there's a way to do that without having to modify ao_set_paused itself.